readme.txt
=== The City Plaza ===
Contributors: weshays
Tags: city, thecity, onthecity, plaza, acs
Requires at least: 2.8
Tested up to: 3.1
Stable tag: 0.8.5

Pull church needs, topics, events, prayers and albums from the Plaza so that you can use them within your WordPress site.


== Description ==

It is possible to pull church needs, topics, events, prayers and albums from the Plaza so that you can use them within your WordPress site.

Data pulled from The City is by default cached for 1 day (24 hours) so that subsequent requests are faster.  This is configurable should your needs be more frequent.


== Installation ==

Install from within WordPress:
<br>
Go to Plugins => Add New => search for "the city plaza".


Install manually:

1. Unzip the file you downloaded.

2. Upload files to a 'the-city-plaza' folder in the '/wp-content/plugins/' directory.

3. Log into your Wordpress admin and activate the plugin through the 'Plugins' menu.

4. In the widgets screen drag the widget to the content section you want it to show up in.

5. Set the fields as shown and then save.



== Frequently Asked Questions == 

= How do I get involved =

Visit our GitHub page and fork the project.
<br>
https://github.com/thecity/thecity-plaza-wordpress


= I found a bug =

Visit this projects GitHub page create an issue for the bug.
<br>
https://github.com/thecity/thecity-plaza-wordpress/issues


= I have a feature request =

Visit this projects GitHub page create an issue for the feature request.
<br>
https://github.com/thecity/thecity-plaza-wordpress/issues


== Screenshots ==

Here is the link to the description of the plugin:
http://developer.onthecity.org/thecity-plugins/wordpress/


== Upgrade Notice == 

None yet...


== Changelog ==

= 0.8.6 =
* Ensure SSL on connection to The City.

= 0.8.5 =
* Added option to make the widget title link back to the Plaza.

= 0.8.4 =
* Missed changing the order of priority for curl and fopen for the plugin. 0.8.3 only updated the thecity-plaza-php library.

= 0.8.3 =
* Changed the order of priority for curl and fopen.  Priority is now given to curl.
* Updated the PlazaPHP lib to the latest.

= 0.8.2 =
* Updated the check for allow_url_fopen.
* Updated the PlazaPHP lib to the latest.

= 0.8.1 =
* Changed where and when the cache table is created/rebuilt.  Now caching to the database really works.

= 0.8.0 =
* Updating caching to use the database instead of the file system.

= 0.7.1 =
* Fixed how cache key was generated.
* Updated how the URL was generated for pulling plaza data.

= 0.7.0 = 
* Added a field for group nickname to pull plaza items specific to that group.

= 0.6.1 =
* Added checks for different ways of getting data from the plaza.  If allow_url_fopen is disabled then an attempt will be made to use cURL.  If both are not available then an error message will be displayed.

= 0.6 =
* Added checkbox to clear the cache when the widget is saved.

= 0.5 =
* Added option to show all plaza items in the same list, which will show a mixture of items depending on the date created (or date "of" if item is an event).  The total number of items in the list will be limited to the number selected.
* Added option to display Plaza item type above title.  If date is also displayed the Plaza item type will show to the right of the date.  This was really meant to be used with the "Show All" option.

= 0.4 =
* Added option for number of items to display.  Up 15 items.

= 0.3 =
* Changed example link under subdomain field to reference OnTheCity.org.
* Moved dates to show before the link.
* Updated styling.
* Using a list instead of div tags.
* Events now show the data of the event rather then the date the event was created.

= 0.2 =
* Added date to list if available.
* Updated CSS to style the date.
* Added option to show dates item was posted.

= 0.1 =
* Initial release of the Plaza-Wordpress plugin.


changelog
0.8.6
* Use SSL connection

0.8.5
* Added option to make the widget title link back to the Plaza.

0.8.4
* Missed changing the order of priority for curl and fopen for the plugin. 0.8.3 only updated the thecity-plaza-php library.

0.8.3
* Changed the order of priority for curl and fopen.  Priority is now given to curl.
* Updated the PlazaPHP lib to the latest.

0.8.2
* Updated the check for allow_url_fopen.
* Updated the PlazaPHP lib to the latest.

0.8.1
* Changed where and when the cache table is created/rebuilt.  Now caching to the database really works.

0.8.0
* Updating caching to use the database instead of the file system.

0.7.1
* Fixed how cache key was generated.
* Updated how the URL was generated for pulling plaza data.

0.7.0
* Added a field for group nickname to pull plaza items specific to that group.

0.6.1
* Added checks for different ways of getting data from the plaza.  If allow_url_fopen is disabled then an attempt will be made to use cURL.  If both are not available then an error message will be displayed.

0.6
* Added checkbox to clear the cache when the widget is saved.

0.5
* Added option to show all plaza items in the same list, which will show a mixture of items depending on the date created (or date "of" if item is an event).  The total number of items in the list will be limited to the number selected.
* Added option to display Plaza item type above title.  If date is also displayed the Plaza item type will show to the right of the date.  This was really meant to be used with the "Show All" option.

0.4
* Added option for number of items to display.  Up 15 items.

0.3
* Changed example link under subdomain field to reference OnTheCity.org.
* Moved dates to show before the link.
* Updated styling.
* Using a list instead of div tags.
* Events now show the data of the event rather then the date the event was created.

0.2
* Added date to list if available.
* Updated CSS to style the date.
* Added option to show dates item was posted.

0.1
* Initial release of the Plaza-Wordpress plugin.

load_data.php
<?php

  require_once '../../../wp-blog-header.php';
  require_once 'lib/plaza-php/lib/the_city.php';
  require_once 'lib/plaza_collection.php';
  require_once 'lib/plaza_wordpress_cache.php';
  
  if( empty( $_GET['subdomain_key'] ) ) {
    echo 'Subdomain not set';
    
    
  } else {

    global $wpdb;
    $cacher = new PlazaWordPressCache( $_GET['subdomain_key'] );
    $cacher->set_db_connection($wpdb);

    $the_city = new TheCity( $_GET['subdomain_key'], true, $cacher);  
    $the_city->add_url_params('wp=1');
    if( isset($_GET['group_nickname']) ) { $the_city->set_group_nickname($_GET['group_nickname']); }
    

    $subdomain_key = $_GET['subdomain_key'];
    $plaza_choice = $_GET['plaza_display'];    
    $items_to_display = $_GET['items_to_display'];    
    $show_dates = isset($_GET['show_dates']) ? $_GET['show_dates'] : '';  
    $show_type = isset($_GET['show_type']) ? $_GET['show_type'] : '';   
    $plaza_choice_key = '';
    $plaza_display = '';

    switch($_GET['plaza_display']) {
      case 'all':
        $plaza_display = new PlazaCollection($the_city, $items_to_display);
        break;
      case 'topics':
        $plaza_display = $the_city->topics($items_to_display); 
        break;
      case 'events':
        $plaza_display = $the_city->events($items_to_display); 
        break;
      case 'prayers':
        $plaza_display = $the_city->prayers($items_to_display); 
        break;
      case 'needs':
        $plaza_display = $the_city->needs($items_to_display); 
        break;
      case 'albums':
        $plaza_display = $the_city->albums($items_to_display); 
        break;
      default:
        $plaza_choice = 'topics';
        $plaza_display = $the_city->topics($items_to_display); 
    }
    
    $html = array();


    $plaza_titles = $plaza_display->titles();
    if( empty($plaza_titles) ) {
      $html[] = "No $plaza_choice found";      
    } else {
      foreach($plaza_titles as $indx => $title) {  
        $item = $plaza_display->select($indx);  

        if($plaza_choice == 'all') {
          $str = get_class($item);
          $item_type_path  = strtolower($str) . 's';
        } else {
          $item_type_path = $plaza_choice;
        }


        $plaza_link_base = 'https://'.$_GET['subdomain_key'].'.onthecity.org/plaza/'.$item_type_path.'/'; 
        $plaza_link = $plaza_link_base . $item->id();   
        $item_date = '';

        if(!empty($show_dates)) {
          $item_created_at = get_class($item) == 'Event' ? $item->starting_at() : $item->created_at();
          if( !empty($item_created_at) ) {
            $item_created_at = date_parse($item_created_at);
            $item_date = implode( array($item_created_at['month'], $item_created_at['day'], $item_created_at['year']), '-');
          }
        }       

        if(!empty($show_type)) {
          $item_type = get_class($item);
          if(empty($item_date)) {
            $item_date = $item_type;
          } else {
            $item_date .= ' :: ' . $item_type;
          }
        }  

        $item_display_date = empty($item_date) ? '' : '<div class="tc_wp_date">' . $item_date . '</div>';
        $html[] = "<li class='tc_wp_item'>$item_display_date<a class='tc_wp_link' href='$plaza_link' target='_blank'>$title</a></li>";
      }
    }

    echo implode($html, '');
  }

?>


widget_info.php
<ul class="tc_wp_content">
  <?php 
    $show_dates = $show_dates == 'on' ? $show_dates : 0;
    $show_type = $show_type == 'on' ? $show_type : 0;

    $otc_params = array();
    $otc_params[] = "subdomain_key=$subdomain_key";
    if(!empty($group_nickname)) { $otc_params[] = "group_nickname=$group_nickname"; }
    $otc_params[] = "plaza_display=$plaza_display";
    $otc_params[] = "items_to_display=$items_to_display";  
    $otc_params[] = "show_dates=$show_dates";  
    $otc_params[] = "show_type=$show_type";

    $url = site_url() . '/wp-content/plugins/the-city-plaza/load_data.php?'.implode('&', $otc_params); 


    if( function_exists('curl_version') ) {
      echo file_get_plaza_contents_with_curl($url);
    } else if( ini_get('allow_url_fopen') == 1 ) { // On
      echo file_get_contents($url);      
    } else {
      echo 'Cannot pull data from plaza.  Either enable allow_url_fopen in the php.ini file, or install curl for php.';
    }
  ?>
</ul>

plaza_scripts.php
<?php

function plaza_wordpress_scripts() {  
  wp_register_script('plaza_wordpress_js', plugins_url( '/the-city-plaza/scripts/plaza_wordpress.js'));   
  wp_enqueue_script('plaza_wordpress_js');  
}

add_action('wp_enqueue_scripts', 'plaza_wordpress_scripts');  



function plaza_wordpress_styles() {  
  wp_register_style('plaza_wordpress_style', plugins_url( '/the-city-plaza/scripts/plaza_wordpress.css'));   
  wp_enqueue_style('plaza_wordpress_style');  
}

add_action('wp_enqueue_scripts', 'plaza_wordpress_styles');  



function file_get_plaza_contents_with_curl($url) {
  $ch = curl_init();
   
  curl_setopt($ch, CURLOPT_HEADER, 0);
  curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1); //Set curl to return the data instead of printing it to the browser.
  curl_setopt($ch, CURLOPT_URL, $url);
   
  $data = curl_exec($ch);
  curl_close($ch);
  return $data;
}


function delete_plaza_cache_directory($dir) {
  // Assume cache has already been cleared.
  if(!file_exists($dir)) { return true; }

  if ($handle = opendir($dir)) {
    $array = array();

    while (false !== ($file = readdir($handle))) {
      if ($file != "." && $file != "..") {

        if(is_dir($dir.$file)) {
          if(!@rmdir($dir.$file)) { // Empty directory? Remove it
            delete_plaza_cache_directory($dir.$file.'/'); // Not empty? Delete the files inside it
          }
        } else {
          @unlink($dir.$file);
        }
      }
    }
    closedir($handle);
    @rmdir($dir);
  }
}

?>

onthecity_plaza.php
<?php
/*
Plugin Name: The City Plaza Widget
Plugin URI: http://thecity.github.com
Description: This widget allows you to pull your OnTheCity.org plaza information into your WordPress website.
Author: Wes Hays
Version: 0.1
Author URI: http://www.OnTheCity.com
*/

include_once 'plaza_scripts.php';


class The_City_Plaza_Widget extends WP_Widget {
  
  function __construct() {
    $widget_ops = array('classname' => 'widget_onthecity_plaza', 
                        'description' => 'Pulls information from your church\'s OnTheCity.org Plaza page.' );
    $this->WP_Widget('the-city-plaza-widget', 'The City Plaza', $widget_ops);
  }
  

  function form($instance) {
    /* Set up some default widget settings. */
		$defaults = array( 'subdomain_key' => '', 'plaza_display' => 'prayers', 'cache_duration' => '86400');
		$instance = wp_parse_args( (array) $instance, $defaults );    
    $title = strip_tags($instance['title']);
    $subdomain_key = strip_tags($instance['subdomain_key']);
    $plaza_display = strip_tags($instance['plaza_display']);
    $cache_duration = strip_tags($instance['cache_duration']);
    ?>

    <p>
      <label for="<?php echo $this->get_field_id('title'); ?>">
        Widget Title: 
        <input class="widefat" 
               id="<?php echo $this->get_field_id('title'); ?>" 
               name="<?php echo $this->get_field_name('title'); ?>" 
               type="text" 
               value="<?php echo attribute_escape($title); ?>" />
       </label>
       <i>The title to display at the top of the widget</i>
    </p>


   <p>
     <label for="<?php echo $this->get_field_id('subdomain_key'); ?>">
       Subdomain: 
       <input class="widefat" 
              id="<?php echo $this->get_field_id('subdomain_key'); ?>" 
              name="<?php echo $this->get_field_name('subdomain_key'); ?>" 
              type="text" 
              value="<?php echo attribute_escape($subdomain_key); ?>" />
      </label>
      <i>Ex: http://[subdomain].MyChurch.org</i>
    </p>

  
    <p>
      <?php 
        $topics_s = $events_s = $prayers_s = $needs_s = $album_s = '';
        switch($instance['plaza_display']) {
          case 'topics':
            $topics_s = 'selected="selected"'; 
            break;
          case 'events':
            $events_s = 'selected="selected"'; 
            break;
          case 'prayers':
            $prayers_s = 'selected="selected"';
            break;
          case 'needs':
            $needs_s = 'selected="selected"'; 
            break;
          case 'albums':
            $album_s = 'selected="selected"';
            break;
        }
      ?> 
    
      <label for="<?php echo $this->get_field_id('plaza_display'); ?>">
        Display:        			
        <select class="widefat" 
                id="<?php echo $this->get_field_id('plaza_display'); ?>" 
                name="<?php echo $this->get_field_name('plaza_display'); ?>">
        		<option value="topics" <?php echo $topics_s; ?> >Topics</option>
        		<option value="events" <?php echo $events_s; ?> >Events</option>
        		<option value="prayers" <?php echo $prayers_s; ?> >Prayers</option>
        		<option value="needs" <?php echo $needs_s; ?> >Needs</option>
        		<option value="albums" <?php echo $album_s; ?> >Albums</option>
        </select>
      </label>
    </p>
    
    
    <p>
      <?php 
        $one_hour = $one_day = $one_week = $one_month = '';
        switch($instance['cache_duration']) {
          case '3600': // One Hour
            $one_hour = 'selected="selected"'; 
            break;
          case '86400': // One Day
            $one_day = 'selected="selected"'; 
            break;
          case '604800': // One Week
            $one_week = 'selected="selected"';
            break;
          case '2592000': // One Month (30 days)
            $one_month = 'selected="selected"'; 
        }
      ?> 
    
      <label for="<?php echo $this->get_field_id('cache_duration'); ?>">
        Cache data for:        			
        <select class="widefat" 
                id="<?php echo $this->get_field_id('cache_duration'); ?>" 
                name="<?php echo $this->get_field_name('cache_duration'); ?>">
        		<option value="3600" <?php echo $one_hour; ?> >One Hour</option>
        		<option value="86400" <?php echo $one_day; ?> >One Day</option>
        		<option value="604800" <?php echo $one_week; ?> >One Week</option>
        		<option value="2592000" <?php echo $one_month; ?> >One Month (30 days)</option>
        </select>
      </label>
    </p>
    <?php
  }
  


  function update($new_instance, $old_instance) {
    $instance = $old_instance;
    $instance['title'] = strip_tags($new_instance['title']);
    $instance['subdomain_key'] = strip_tags($new_instance['subdomain_key']);
    $instance['plaza_display'] = strip_tags($new_instance['plaza_display']);
    $instance['cache_duration'] = strip_tags($new_instance['cache_duration']);
    return $instance;
  }
  


  function widget($args, $instance) {
    extract($args);

    $title = empty($instance['title']) ? ' ' : apply_filters('widget_title', $instance['title']);
    $subdomain_key = empty($instance['subdomain_key']) ? ' ' : $instance['subdomain_key'];
    $plaza_display = empty($instance['plaza_display']) ? ' ' : $instance['plaza_display'];
    $cache_duration = empty($instance['cache_duration']) ? ' ' : $instance['cache_duration'];


    echo $before_widget;
    if (!empty( $title )) {
        echo $before_title . $title . $after_title;
    };

    include dirname(__FILE__).'/widget_info.php';

    echo $after_widget;
  }
  
}

add_action('widgets_init', create_function('', 'return register_widget("The_City_Plaza_Widget");'));

?>

lib/plaza_collection.php
<?php


  /**
   * Set the default Time Zone if not already set for the user.
   * I don't think this will actually be used anywhere and is 
   * really only for PHP 5.3.
   */

  if (!ini_get('date.timezone')) {
    date_default_timezone_set('America/Los_Angeles');
  }


  /** 
   * This class is a wrapper for the topics page.
   *
   * @package TheCity
   */
  class PlazaCollection extends Plaza {

    // Holds the data.
    private $plaza_data = array();

    /**
     *  Constructor.
     *
     * @param TheCity $the_city An initialized TheCity object.
     */
    public function __construct($the_city, $items_to_display) {

      $topics = $the_city->topics($items_to_display); 
      $events = $the_city->events($items_to_display); 
      $prayers = $the_city->prayers($items_to_display); 
      $needs = $the_city->needs($items_to_display); 
      $albums = $the_city->albums($items_to_display);       

      $temp = array();

      $i = 0;
      while( ($topic = $topics->select($i)) != NULL ) { 
        $key = $this->_build_time_key( $topic->created_at() );
        $temp[$key] = $topic;
        $i++;
      } 

      $i = 0;
      while( ($event = $events->select($i)) != NULL ) { 
        $key = $this->_build_time_key( $event->starting_at() );
        $temp[$key] = $event;
        $i++;
      }      

      $i = 0;
      while( ($prayer = $prayers->select($i)) != NULL ) { 
        $key = $this->_build_time_key( $prayer->created_at() );
        $temp[$key] = $prayer;
        $i++;
      }    

      $i = 0;
      while( ($need = $needs->select($i)) != NULL ) { 
        $key = $this->_build_time_key( $need->created_at() );
        $temp[$key] = $need;
        $i++;
      }  

      $i = 0;
      while( ($album = $albums->select($i)) != NULL ) { 
        $key = $this->_build_time_key( $album->created_at() );
        $temp[$key] = $album;
        $i++;
      }

      krsort($temp); // Sort by keys from high to low

      $this->plaza_data = array_slice($temp, 0, $items_to_display);
    }
    


    /**
     *  All the Plaza item titles.
     *
     *  @return array of titles.
     */
    public function all_titles() {
      $items = array();
      foreach ($this->plaza_data as $item) { $items[] = $item->title(); }
      return $items;
    }
    
    
    /**
     * Get the specified Plaza item.
     *
     * @param index The index of the item to get all the information for.
     *
     * @return Mixed It could be a Topic, Event, Prayer, Need or Album.
     */
    public function select($index) {
      if( !isset($this->plaza_data[$index]) ) { return null; }
      return $this->plaza_data[$index];
    }
    
    

    private function _build_time_key($created_at_str) {
      return strtotime( $created_at_str );
    }

  }
?>

the_city_plaza.php
<?php
/*
Plugin Name: The City Plaza Widget
Plugin URI: http://developer.onthecity.org/thecity-plugins/wordpress/
Description: This widget allows you to pull your OnTheCity.org plaza information into your WordPress website.
Author: Wes Hays
Version: 0.8.6
Author URI: http://www.OnTheCity.org
*/

include_once 'plaza_scripts.php';
require_once 'lib/plaza-php/lib/the_city.php';
require_once 'lib/plaza_wordpress_cache.php';


class The_City_Plaza_Widget extends WP_Widget {
  
  function __construct() {
    $widget_ops = array('classname' => 'the_city_plaza_widget', 
                        'description' => 'Pulls information from your church\'s OnTheCity.org Plaza page.' );
    $this->WP_Widget('the-city-plaza-widget', 'The City Plaza', $widget_ops);
  }
  

  function form($instance) {
    /* Set up some default widget settings. */
		$defaults = array( 'subdomain_key' => '', 
                       'title_link' => '1',
                       'group_nickname' => '',
                       'plaza_display' => 'prayers', 
                       'items_to_display' => '10',
                       'show_dates' => '0', 
                       'show_type' => '0',
                       'cache_duration' => '86400');

		$instance = wp_parse_args( (array) $instance, $defaults );    

    $title = strip_tags($instance['title']);
    $subdomain_key = strip_tags($instance['subdomain_key']);
    $title_link = strip_tags($instance['title_link']);
    $group_nickname = strip_tags($instance['group_nickname']);
    $plaza_display = strip_tags($instance['plaza_display']);
    $items_to_display = strip_tags($instance['items_to_display']);
    $show_dates = strip_tags($instance['show_dates']);
    $show_type = strip_tags($instance['show_type']);
    $cache_duration = strip_tags($instance['cache_duration']);

    $clear_cache_msg = '';
    if (isset($_POST['clear_cache_now'])) {
      global $wpdb;
      $cacher = new PlazaWordPressCache( $subdomain_key );
      $cacher->set_db_connection($wpdb);
      $cacher->clear_cache();
      $clear_cache_msg = 'Cache cleared';
    }

    ?>

    <p>
      <label for="<?php echo $this->get_field_id('title'); ?>">
        Widget Title: 
        <input class="widefat" 
               id="<?php echo $this->get_field_id('title'); ?>" 
               name="<?php echo $this->get_field_name('title'); ?>" 
               type="text" 
               value="<?php echo attribute_escape($title); ?>" />
       </label>
       <i>The title to display at the top of the widget</i>

      <br>

      <?php
        $title_link_checked = empty($title_link) ? '' : 'checked="checked"';
      ?>
      <label for="<?php echo $this->get_field_id('title_link'); ?>">
        <input type="checkbox" 
               id="<?php echo $this->get_field_id('title_link'); ?>" 
               name="<?php echo $this->get_field_name('title_link'); ?>"
               <?php echo $title_link_checked ?> /> Title links to main plaza page      
      </label>         
    </p>


    <p>
     <label for="<?php echo $this->get_field_id('subdomain_key'); ?>">
       Subdomain: 
       <input class="widefat" 
              id="<?php echo $this->get_field_id('subdomain_key'); ?>" 
              name="<?php echo $this->get_field_name('subdomain_key'); ?>" 
              type="text" 
              value="<?php echo attribute_escape($subdomain_key); ?>" />
      </label>
      <i>Ex: https://[subdomain].OnTheCity.org</i>
    </p>


   <p>
     <label for="<?php echo $this->get_field_id('group_nickname'); ?>">
       Group Nickname (optional): 
       <input class="widefat" 
              id="<?php echo $this->get_field_id('group_nickname'); ?>" 
              name="<?php echo $this->get_field_name('group_nickname'); ?>" 
              type="text" 
              value="<?php echo attribute_escape($group_nickname); ?>" />
      </label>
      <i>Only items for this group will be pulled.</i>
    </p>    

  
    <p>
      <?php 
        $topics_s = $events_s = $prayers_s = $needs_s = $album_s = '';
        switch($instance['plaza_display']) {
          case 'all':
            $all_s = 'selected="selected"'; 
            break;
          case 'topics':
            $topics_s = 'selected="selected"'; 
            break;
          case 'events':
            $events_s = 'selected="selected"'; 
            break;
          case 'prayers':
            $prayers_s = 'selected="selected"';
            break;
          case 'needs':
            $needs_s = 'selected="selected"'; 
            break;
          case 'albums':
            $album_s = 'selected="selected"';
            break;
        }
      ?> 
    
      <label for="<?php echo $this->get_field_id('plaza_display'); ?>">
        Display:        			
        <select class="widefat" 
                id="<?php echo $this->get_field_id('plaza_display'); ?>" 
                name="<?php echo $this->get_field_name('plaza_display'); ?>">
            <option value="all" <?php echo $all_s; ?> >Show All</option>
        		<option value="topics" <?php echo $topics_s; ?> >Topics</option>
        		<option value="events" <?php echo $events_s; ?> >Events</option>
        		<option value="prayers" <?php echo $prayers_s; ?> >Prayers</option>
        		<option value="needs" <?php echo $needs_s; ?> >Needs</option>
        		<option value="albums" <?php echo $album_s; ?> >Albums</option>
        </select>
      </label>


      <?php
        $show_dates_checked = empty($show_dates) ? '' : 'checked="checked"';
      ?>
      <label for="<?php echo $this->get_field_id('show_dates'); ?>">
        <input type="checkbox" 
               id="<?php echo $this->get_field_id('show_dates'); ?>" 
               name="<?php echo $this->get_field_name('show_dates'); ?>"
               <?php echo $show_dates_checked ?> /> Show Dates        
      </label>

      <br>

      <?php
        $show_type_checked = empty($show_type) ? '' : 'checked="checked"';
      ?>
      <label for="<?php echo $this->get_field_id('show_type'); ?>">
        <input type="checkbox" 
               id="<?php echo $this->get_field_id('show_type'); ?>" 
               name="<?php echo $this->get_field_name('show_type'); ?>"
               <?php echo $show_type_checked ?> /> Show Plaza item type above title      
      </label>      
    </p>
    

    <p>
      <label for="<?php echo $this->get_field_id('items_to_display'); ?>">
        Items to display:              
        <select class="widefat" 
                id="<?php echo $this->get_field_id('items_to_display'); ?>" 
                name="<?php echo $this->get_field_name('items_to_display'); ?>">

            <?php 
              $item_count_selected = '';
              for($i=1; $i<=15; $i++) {
                $item_count_selected = $items_to_display == $i ? 'selected="selected"' : '';
                echo "<option value=\"$i\" $item_count_selected>$i</option>";                
              }
            ?>
        </select>
      </label>      
    </p>

    
    <p>
      <?php 
        $one_hour = $one_day = $one_week = $one_month = '';
        switch($instance['cache_duration']) {
          case '3600': // One Hour
            $one_hour = 'selected="selected"'; 
            break;
          case '86400': // One Day
            $one_day = 'selected="selected"'; 
            break;
          case '604800': // One Week
            $one_week = 'selected="selected"';
            break;
          case '2592000': // One Month (30 days)
            $one_month = 'selected="selected"'; 
        }
      ?> 
    
      <label for="<?php echo $this->get_field_id('cache_duration'); ?>">
        Cache data for:        			
        <select class="widefat" 
                id="<?php echo $this->get_field_id('cache_duration'); ?>" 
                name="<?php echo $this->get_field_name('cache_duration'); ?>">
        		<option value="3600" <?php echo $one_hour; ?> >One Hour</option>
        		<option value="86400" <?php echo $one_day; ?> >One Day</option>
        		<option value="604800" <?php echo $one_week; ?> >One Week</option>
        		<option value="2592000" <?php echo $one_month; ?> >One Month (30 days)</option>
        </select>
      </label>

      <label for="clear_cache_now">
        <input type="checkbox" id="clear_cache_now" name="clear_cache_now" /> Clear cache on save       
      </label>

      <?php if(!empty($clear_cache_msg)) { echo '<div style="color:#990000">'.$clear_cache_msg.'</div>'; } ?>
    </p>
    <?php
  }



  function update($new_instance, $old_instance) {
    $instance = $old_instance;

    // This is done this way for older versions of PHP otherwise I would combine them all into one line.
    $nickname = strip_tags($new_instance['group_nickname']);
    $nickname = strtolower($nickname);
    $nickname = trim($nickname);

    $instance['title'] = strip_tags($new_instance['title']);
    $instance['title_link'] = strip_tags($new_instance['title_link']);
    $instance['subdomain_key'] = strip_tags($new_instance['subdomain_key']);
    $instance['group_nickname'] = $nickname;
    $instance['plaza_display'] = strip_tags($new_instance['plaza_display']);
    $instance['items_to_display'] = strip_tags($new_instance['items_to_display']);
    $instance['show_dates'] = strip_tags($new_instance['show_dates']);
    $instance['show_type'] = strip_tags($new_instance['show_type']);
    $instance['cache_duration'] = strip_tags($new_instance['cache_duration']);
    return $instance;
  }
  


  function widget($args, $instance) {
    extract($args);

    $title = empty($instance['title']) ? ' ' : apply_filters('widget_title', $instance['title']);
    $title_link = empty($instance['title_link']) ? '' : $instance['title_link'];
    $subdomain_key = empty($instance['subdomain_key']) ? ' ' : $instance['subdomain_key'];
    $group_nickname = empty($instance['group_nickname']) ? '' : $instance['group_nickname'];
    $plaza_display = empty($instance['plaza_display']) ? ' ' : $instance['plaza_display'];
    $items_to_display = empty($instance['items_to_display']) ? '10' : $instance['items_to_display'];
    $show_dates = empty($instance['show_dates']) ? ' ' : $instance['show_dates'];
    $show_type = empty($instance['show_type']) ? ' ' : $instance['show_type'];
    $cache_duration = empty($instance['cache_duration']) ? ' ' : $instance['cache_duration'];

    echo $before_widget;
    if (!empty( $title )) {
      $display_title = $title;
      if(!empty($title_link)) {
        $display_title = '<a href="http://'.$subdomain_key.'.onthecity.org/plaza" target="_blank">'.$title.'</a>';
      }
      echo $before_title . $display_title . $after_title;
    };

    include dirname(__FILE__).'/widget_info.php';

    echo $after_widget;
  }
  
}

add_action('widgets_init', create_function('', 'return register_widget("The_City_Plaza_Widget");'));

?>